File: cs50x/pset6/vigenere/vigenere.py
import sys
from cs50 import get_string

# Make sure script is ran with appropriate number of args

if len(sys.argv) != 2:
    sys.stdout.write("Please provide key.\n")
    sys.exit(1)

# prep to work with input

input_key = sys.argv[1]

# strings are immutable in python, so work with lists

cipher_key = list()

# Complain if any non-alphabetical letter

if not input_key.isalpha():
    sys.stdout.write("Please input only letters A-Z, a-z.\n")
    sys.exit(1)

# Make cipher key

for c in input_key:
    if ord(c) > 64 and ord(c) < 91:
        cipher_key.append(ord(c) - 65)
    else:
        cipher_key.append(ord(c) - 97)

# set up counter

counter = 0

# get user input for plain text

plain_text = get_string("plaintext: ")
print("ciphertext: ", end="")

# print cipher text

for c in plain_text:
    char = ord(c)
    if counter == len(cipher_key):
        counter = 0
    if char > 64 and char < 91:
        distance = 90 - char
        key_num = cipher_key[counter] % 26
        if key_num < distance or key_num == distance:
            print(chr(char + key_num), end="")
        else:
            print(chr(64 + key_num - distance), end="")
        counter += 1
    elif char > 96 and char < 123:
        distance = 122 - char
        key_num = cipher_key[counter] % 26
        if key_num < distance or key_num == distance:
            print(chr(char + key_num), end="")
        else:
            print(chr(96 + key_num - distance), end="")
        counter += 1
    else:
        print(f"{c}", end="")

print("")

File: cs50x/pset2/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

void RotateUpperCase(char c, int key);
void RotateLowerCase(char c, int key);

int main(int argc, string argv[])
{
    if (argc == 1 || argc > 2) {
        fprintf(stdout, "Please provide key");
        return 1;
    }

    // Key
    char *key = argv[1];
    size_t key_len = strlen(key);

    // Complain if any non-alphabetical letter
    for (int i = 0; i < key_len; i++) {
        if (!isalpha(key[i])) {
            fprintf(stdout, "Please input only letters A-Z, a-z.");
            return 1;
        }
    };

    // Make cipher key
    for (int i = 0; i < key_len; i++) {
        if (key[i] > 64 && key[i] < 91) {
            // Uppercase
            key[i] -= 65;
        } else {
            // Lowercase
            key[i] -= 97;
        }
    };

    int counter = 0;

    // get ciphertext
    string cipher_text = GetString();
    size_t text_len = strlen(cipher_text);

    for (int i = 0; i < text_len; i++) {
        if (counter == key_len) {
            counter = 0;
        }
        if (cipher_text[i] > 64 && cipher_text[i] < 91) {
            RotateUpperCase(cipher_text[i], key[counter] % 26);
            counter++;
        } else if (cipher_text[i] > 96 && cipher_text[i] < 123) {
            RotateLowerCase(cipher_text[i], key[counter] % 26);
            counter++;
        } else {
            printf("%c", cipher_text[i]);
        }
    }
    printf("\n");
    return 0;
}

void RotateUpperCase(char c, int key) {
    int distance = 90 - c;
    if (key < distance || key == distance) {
        printf("%c", c + key);
    } else {
        printf("%c", 64 + (key - distance));
    }
}

void RotateLowerCase(char c, int key) {
    int distance = 122 - c;
    if (key < distance || key == distance) {
        printf("%c", c + key);
    } else {
        printf("%c", 96 + (key - distance));
    }
}